simplefs/src/sb.rs

```diff
@@ -0,0 +1,87 @@
+use zerocopy::{AsBytes, FromBytes};
+
+/// The first block of the file system storing information critical for mounting
+/// the file system and verifying the underlying disk is formatted correctly.
+///
+/// Keeps the size of the file system by tracking the number of blocks allocated
+/// to the inode and data block groups. The number of nodes available in the filesystem
+/// ultimately sets the upper bound on how many files can exist.
+///
+/// Some files, such as files with no data and symbolic links don't allocate any
+/// data blocks but do allocate inode blocks.
+#[repr(C)]
+#[derive(Debug, PartialEq, AsBytes, FromBytes, Clone, Copy)]
+pub struct SuperBlock {
+    /// A 32-bit identifying string, in this case SFSB.
+    pub sb_magic: u32,
+    /// Assuming 256 bytes per inode a 4K block can hold 16 inodes.
+    pub inodes_count: u32,
+    /// All the remaining blocks are allocating to storing user data.
+    pub blocks_count: u32,
+    /// All blocks currently in use by the filesystem.
+    pub reserved_blocks_count: u32,
+    /// All blocks available to be allocated by the system.
+    pub free_blocks_count: u32,
+    /// The number of remaining available inodes.
+    pub free_inodes_count: u32,
+    /// The index of the next available free block.
+    pub free_list: u32,
+}
+
+impl SuperBlock {
+    pub fn new() -> Self {
+        Self {
+            sb_magic: 0, // Default to invalid zero value.
+            inodes_count: 0,
+            blocks_count: 0,
+            reserved_blocks_count: 0,
+            free_blocks_count: 0,
+            free_inodes_count: 0,
+            free_list: 0,
+        }
+    }
+
+    /// Attempts to parse a buffer as a SuperBlock returning a new owned instance
+    /// of the block. If the block is invalid, calling parse will cause a panic.
+    pub fn parse(buf: &[u8], magic: u32) -> Self {
+        let sb: *const SuperBlock = buf.as_ptr() as *const SuperBlock;
+
+        unsafe {
+            assert_eq!(magic, (*sb).sb_magic, "Superblock magic constant invalid.");
+            SuperBlock::from(*sb)
```

Comment:
there is not a safe approach to do this is there? i can't think of one off hand

Reply:
I couldn't think of one that didn't cause another allocation. I'm not sure it matters _that_ much if we allocate here since we would presumably only do this at start up anyway... we can circle back to this bit (maybe with some benchmarks)